server/src/main/java/org/opensearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -292,6 +291,65 @@ public ShardIterator activeInitializingShardsRankedIt(
         return new PlainShardIterator(shardId, ordered);
     }
 
+    /**
+     * *
+     * @param wrrWeight Weighted round-robin weight entity
+     * @param nodes discovered nodes in the cluster
+     * @return an interator over active and initializing shards, ordered by weighted round-robin
+     * scheduling policy. Making sure that initializing shards are the last to iterate through.
+     */
+    public ShardIterator activeInitializingShardsWRR(WRRWeight wrrWeight, DiscoveryNodes nodes) {
+        final int seed = shuffler.nextSeed();
+        ArrayList<ShardRouting> ordered = new ArrayList<>(activeShards.size() + allInitializingShards.size());
+        ArrayList<ShardRouting> orderedActiveShards = getShardsWRR(activeShards, wrrWeight, nodes);
+        ordered.addAll(shuffler.shuffle(orderedActiveShards, seed));
+        if (!allInitializingShards.isEmpty()) {
+            ArrayList<ShardRouting> orderedInitializingShards = getShardsWRR(allInitializingShards, wrrWeight, nodes);
+            ordered.addAll(shuffler.shuffle(orderedInitializingShards, seed));
+        }
+        return new PlainShardIterator(shardId, ordered);
+    }
+
+    /**
+     *
+     * @param shards shards to be ordered using weighted round-robin scheduling policy
+     * @param wrrWeight weights to be considered for routing
+     * @param nodes discovered nodes in the cluster
+     * @return list of shards ordered using weighted round-robin scheduling.
+     */
+    private ArrayList<ShardRouting> getShardsWRR(List<ShardRouting> shards, WRRWeight wrrWeight, DiscoveryNodes nodes) {
+        List<WeightedRoundRobin.Entity<ShardRouting>> weightedShards = calculateShardWeight(shards, wrrWeight, nodes);
+        WeightedRoundRobin<ShardRouting> wrr = new WeightedRoundRobin<>(weightedShards);
+        List<WeightedRoundRobin.Entity<ShardRouting>> wrrOrderedActiveShards = wrr.orderEntities();
+        ArrayList<ShardRouting> orderedActiveShards = new ArrayList<>(activeShards.size());
+        for (WeightedRoundRobin.Entity<ShardRouting> shardRouting : wrrOrderedActiveShards) {
+            orderedActiveShards.add(shardRouting.getTarget());
+        }
+        return orderedActiveShards;
+    }
```

Comment:
Can we cache post a cluster state update rather than always computing this

Reply:
Should we do a lazy calculation ie whenever a request for the shard comes, we check in cache and if its marked dirty do calculate it. If we do calculate on every cluster state update, then it will trigger WRR routing calculation for all the shards in the cluster which can be an additional overhead.

Reply:
Added caching layer